tomoya-sforzando/etude-Prefect
# feat: build base of abstraction of flow

## Describe the PR

Build base of abstraction of flow.

## Screenshots

|達成したいタスクを記述した場合は、そのタスクに必要なタスクだけでFlowを構築する|記述なしの場合は、全てを達成するFlowを構築する|
|---|---|
|<img width="937" alt="Screen Shot 2022-04-12 at 19 30 39" src="https://user-images.githubusercontent.com/40506652/162941853-93985c3a-d704-4ad4-a214-dad967ebd458.png">|<img width="937" alt="Screen Shot 2022-04-12 at 19 31 46" src="https://user-images.githubusercontent.com/40506652/162941877-25a2b40a-66de-40c5-b37e-9cd5f40be91e.png">|
|![Screen Shot 2022-04-12 at 19 32 40](https://user-images.githubusercontent.com/40506652/162942102-9f6b97bf-9adc-4e99-b6b8-514cacda33b1.png)|![Screen Shot 2022-04-12 at 19 32 22](https://user-images.githubusercontent.com/40506652/162941971-5f15e4e6-9fb2-4f39-b1ca-db6fc75ed6bc.png)

## Detail of the change

- Add `flows/abstract_flow.py`
- Add `flows/abstract_settings.py`
- Add `flows/idetail/idetail_flow.py`
- Add `flows/idetail/idetail_settings.py`
- Add `tasks/` as sample Task

## Anticipated impacts

None

## To reproduce

Steps to check the behavior:

1. `make up`
1. Open `http://localhost:8080/`
1. Check Run histories

## Additional context

- Flowで組み立てられるTaskの依然関係は、 `build_basic_flow()` に記述する
    - `Task.set_upstream()` を使って書くと、複数の依存関係がある場合、また、それが後に増えた場合、 単純に `Task.set_upstream()` での記述を増やせば良いだけなので、シンプルに依存関係を列挙できる。ただし、依存関係が多くなると行数が増えるので読みづらくなっていく
        - https://github.com/tomoya-sforzando/etude-Prefect/blob/733d050c7e5bf83a11e2570c430dcb2096d6eff7/flow/flows/idetail/idetail_flow.py#L19
    - `Task.set_dependencies()` を使った場合は、複数の依存関係もまとめて簡潔に書けるメリットがある
        - https://github.com/tomoya-sforzando/etude-Prefect/blob/5668c23909020a5658a272bfd5d2564cc3451996/flow/flows/idetail/idetail_flow.py#L19
    - 好みの問題かもしれないが、 `Task.set_dependencies()` での記述の方が良いかも。
